File: 81.py
# 3244 I nasaite
from time import time
matricha = []
import copy
nnnnn = time()
def razdeli(straka):
    nov = []
    n = ""
    for i in straka:
        n += i
        if i == ",":
            nov.append(int(n[:-1]))
            n = ""
    return nov

f = open("81.txt")
for liniya in f:
    matricha.append(razdeli(liniya))
def sasedi(kord, mat):
    # if kord = (9,9)
    # to (8, 9) (9, 8)  
    st, n = kord[0], kord[1]
    a = [(st-1, n), (st, n-1)]
    nov = []
    for i in a:
        try:
            if i[0] != -1:
                if i[1] != -1:
                    bb = mat[i[0]][i[1]]
                    nov.append(i)
        except IndexError:
            pass
    return nov

def sam_puti(nac,kan):
    puti = [(len(kan)-1,len(kan)-1)]
    a = 0
    while True:
        kord = puti[a] 
        zak = False
        samo_chislo = kan[kord[0]][kord[1]]
        sos = sasedi(kord, kan)
        for i in sos:
            chislo_s = nac[kord[0]][kord[1]]
            if samo_chislo - chislo_s == kan[i[0]][i[1]]:
                puti.append(i)
                if i == (0, 0):
                    zak = True
                break
        if zak:
            break        
        a +=1
    puti.reverse()
    return puti 

#print matricha
def nachni(matricha):
    ma_nov = matricha
    s = 0
    for i in xrange(len(matricha)):
        s += ma_nov[0][i]
        ma_nov[0][i] = s
    s = 0    
    for i in xrange(len(matricha)):
        s += ma_nov[i][0]
        ma_nov[i][0] = s
    return ma_nov

def vt(st,n):
    #428 = (1,1)
    return [st+1,n-1]

def prv_vn(st,n, st1, n1):
    return [st1,n]

#mam = copy.deepcopy(matricha)
def kan(matricha):
    nov = matricha
    for i in xrange(len(matricha)-1):
        for j in xrange(1,len(matricha)):
            # nov NOVAYA
            kord = vt(i,j)
            per = nov[i][j]#804
            vtor = nov[kord[0]][kord[1]]#332
            
            kord_tr = prv_vn(i,j, kord[0], kord[1])
            
            tretiy = matricha[kord_tr[0]][kord_tr[1]]
            min_sribi_p_v = [per,vtor]
            nov[kord_tr[0]][kord_tr[1]] = min(min_sribi_p_v)+tretiy
    return nov
a = nachni(matricha)
ae = kan(a)
print ae[-1][-1], time()-nnnnn

#pp = sam_puti(mam,ae)
#for i in pp:
#    print mam[i[0]][i[1]],

File: 81.java
package javaapplication1;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Scanner;

public class JavaApplication1 {
//"c:\\81.txt"

    public static void main(String[] args) throws FileNotFoundException{

        ArrayList<ArrayList<Integer>> matricha = nachni(fail_v_matrizhu());
        
        ArrayList<ArrayList<Integer>> matricha_nov = new ArrayList<>();
        for (int i = 0; i != matricha.size(); i++){
            ArrayList<Integer> nov;
            nov = (ArrayList<Integer>) matricha.get(i).clone();
            matricha_nov.add(nov);
        }
        for(int i = 0; i != matricha_nov.size()-1; i++){
            for(int j = 1; j != matricha_nov.size(); j++){
                int per = matricha_nov.get(i).get(j);
                int vtor = matricha_nov.get(i+1).get(j-1);
                int tretiy = matricha.get(i+1).get(j);
                int menishi = min(per, vtor);
                matricha_nov.get(i+1).set(j, menishi + tretiy);
            }
            
        }
        for(int i = 0; i != matricha_nov.size(); i++){
            System.out.println(matricha_nov.get(i));
        }
        }
    

    public static Integer min(int a, int b){
        if (a < b){
            return a;
        }
        else if (a > b){
            return b;
        } 
        else{
            return a;
        }
        
    }
    
    public static ArrayList<ArrayList<Integer>> nachni(ArrayList<ArrayList<Integer>> matricha){
        int suma = 0;
        for(int i = 0; i != matricha.size(); i++){
            suma += matricha.get(0).get(i);
            matricha.get(0).set(i, suma);
        }
        suma = 0;
        for(int i = 0; i != matricha.size(); i++){
            suma += matricha.get(i).get(0);
            matricha.get(i).set(0, suma);
        }
        return matricha;
        
    }


    public static ArrayList<Integer> razdeli(String stroka) {
        ArrayList<Integer> nov = new ArrayList<>();
        String n = "";
        for (int i = 0; i != stroka.length(); i++) {
            String i_p = stroka.substring(i, i + 1);
            n += i_p;
            if (",".equals(i_p)) {
                int razmer = n.length();
                int chislo = Integer.parseInt(n.substring(0, razmer - 1));
                nov.add(chislo);
                n = "";
            }
        }
        return nov;
    }

    public static ArrayList<ArrayList<Integer>> fail_v_matrizhu() throws FileNotFoundException {
        String s = "";
        Scanner in = new Scanner(new File("c:\\81.txt"));
        ArrayList<ArrayList<Integer>> matricha = new ArrayList<>();
        while (in.hasNext()) {
            matricha.add(razdeli(in.nextLine()));
        }
        return matricha;
    }
}